fetcher/scion_cppki_verify.go

```
@@ -0,0 +1,249 @@
+		"-inform", "PEM", "-print_certs", "-out", asCertHumanChain).Run()
+	if err != nil {
+		return fmt.Errorf("certificate validation failed: "+
+			"unable to gather included certificates from signature: %w", err)
+	}
+
+	// Split signer and ca certificate
+	rawASCertHumanChain, _ := os.ReadFile(asCertHumanChain)
+	re := regexp.MustCompile("-*?BEGIN CERTIFICATE-*?\\n[\\s\\S]*-*?END CERTIFICATE-*?\\n")
+	matches := re.FindAllString(string(rawASCertHumanChain), -1)
+
+	asCertRaw := []byte(matches[0])
+	rawASCertPem, _ := pem.Decode(asCertRaw)
+	if rawASCertPem != nil {
+		asCertRaw = rawASCertPem.Bytes
+	}
+	cert, err := x509.ParseCertificate(asCertRaw)
+	if err != nil {
+		return fmt.Errorf("certificate validation failed: "+
+			"unable to parse signer certificates from signature: %w", err)
+	}
+
+	// Check signer AS ID matches unverifiedIA
+	certSubjectASID := ""
+	OIDNameIA := asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 55324, 1, 2, 1}
+	for _, dn := range cert.Subject.Names {
+		if dn.Type.Equal(OIDNameIA) {
+			certSubjectASID = dn.Value.(string)
+			break
+		}
+	}
+	if certSubjectASID != unverifiedIA {
+		return fmt.Errorf("certificate validation failed: "+
+			"signer AS certificate subject does not match AS ID included in topology: "+
+			"expected: %s, actual: %s", unverifiedIA, certSubjectASID)
+	}
+
+	// Store certificate chain extracted from signature
+	asCertChainPath := path.Join(bootstrapperPath, "as_cert_chain.pem")
+	rawASCertChain := strings.Join(matches, "\n")
+	_ = os.WriteFile(asCertChainPath, []byte(rawASCertChain), 0666)
+
+	// verify AS certificate chain back to TRC:
+	err = exec.CommandContext(ctx, "scion-pki", "certificate", "verify",
+		"--trc", trustAnchorTRC, asCertChainPath).Run()
+	if err != nil {
+		return fmt.Errorf("certificate validation failed: unable to validate certificate chain: %w", err)
+	}
+
+	// We now have a signed topology with a valid signature and a certificate chain back to a TRC, write out the payload
```

Do we really need to extract the payload again? Just rename `topology.json.unverifed` to `topology.json`?

---

It is not necessary to extract the payload again. The initial signature validation was done to fail early, avoid checking a certificate chain if we already know the signature is not valid.
That we already have extracted the (unverified) payload was an artifact of using the IA in the unverified payload to drive the verification.

The payload verification is now driven by the identity of the signer and the check that the signer identity matches the IA contained in the signed payload is performed as a separate step.